lib/zclient.c

```diff
@@ -3341,3 +3386,83 @@ int32_t zapi_capabilities_decode(struct stream *s, struct zapi_cap *api)
 stream_failure:
 	return 0;
 }
+
+int zclient_send_srv6_function(struct zclient *zclient,
+			       const struct srv6_function *function)
+{
+	struct stream *s;
+
+	s = zclient->obuf;
+	stream_reset(s);
+	zapi_srv6_function_encode(ZEBRA_SRV6_FUNCTION_ADD, s, function);
+	zclient_send_message(zclient);
+	return 0;
+}
+
+int zapi_srv6_locator_encode(uint8_t cmd, struct stream *s,
+			     const struct srv6_locator *locator)
+{
+	uint8_t namelen;
+
+	zclient_create_header(s, cmd, VRF_DEFAULT);
+	namelen = strlen(locator->name);
+	stream_putc(s, namelen);
+	stream_put(s, locator->name, namelen);
+	stream_put(s, &locator->prefix, sizeof(locator->prefix));
+	stream_putc(s, locator->function_bits_length);
+	stream_putw_at(s, 0, stream_get_endp(s));
+	return 0;
+}
+
+int zapi_srv6_locator_decode(struct stream *s, struct srv6_locator *locator)
+{
+	uint8_t namelen;
+
+	STREAM_GETC(s, namelen);
+	stream_get(locator->name, s, namelen);
+	locator->name[namelen] = '\0';
+	stream_get(&locator->prefix, s, sizeof(locator->prefix));
+	locator->function_bits_length = stream_getc(s);
+stream_failure:
+	return 0;
+}
+
+int zapi_srv6_function_encode(uint8_t cmd, struct stream *s,
+			      const struct srv6_function *function)
+{
+	uint8_t namelen;
+
+	zclient_create_header(s, cmd, VRF_DEFAULT);
+	namelen = strlen(function->locator_name);
+	stream_putc(s, namelen);
+	stream_put(s, function->locator_name, namelen);
+	stream_put(s, &function->prefix, sizeof(function->prefix));
+	stream_putc(s, function->owner_proto);
+	stream_putw(s, function->owner_instance);
+	stream_putl(s, function->request_key);
+	stream_putl(s, function->ifindex);
+	stream_putl(s, function->action);
+	stream_write(s, &function->ctx, sizeof(function->ctx));
+	stream_putc(s, function->explicit_allocate);
+	stream_putw_at(s, 0, stream_get_endp(s));
+	return 0;
+}
+
+int zapi_srv6_function_decode(struct stream *s, struct srv6_function *function)
+{
+	uint8_t namelen;
+
+	STREAM_GETC(s, namelen);
+	memset(function->locator_name, 0, sizeof(function->locator_name));
+	stream_get(function->locator_name, s, namelen);
+	stream_get(&function->prefix, s, sizeof(function->prefix));
+	STREAM_GETC(s, function->owner_proto);
+	STREAM_GETW(s, function->owner_instance);
+	STREAM_GETL(s, function->request_key);
+	STREAM_GETL(s, function->ifindex);
+	STREAM_GETL(s, function->action);
+	stream_get(&function->ctx, s, sizeof(function->ctx));
+	STREAM_GETC(s, function->explicit_allocate);
```

Comment:
Don't mix these, use the error checking STREAM_* variants.

Reply:
I fixed it